File: frontend/src/types/transaction.ts
import { util } from '@app/utils/wrappers';
import {
    CreateMoneyTransactionRequest,
    CreateProductTransactionRequest,
} from '@app/codegen/transaction/create';
import {
    ViewHistoryRequest,
    ViewTransactionsResponse,
    TransactionEntry_Status,
    TransactionEntry,
} from '@app/codegen/transaction/history';
import {
    DecideTransactionRequest,
} from '@app/codegen/transaction/transaction';
import {
    ViewCurrentTransactionsRequest,
    CurrentTransactionsResponse,
    CurrentTransactionEntry,
} from '@app/codegen/transaction/unpaid';

export namespace TransactionAPI {
    export interface TransactionRecord {
        transaction_id: number;
        product: string;
        count: number;
        amount: number;
        status: 'created' | 'approved' | 'rejected' | 'completed';
        updated_at: string;
        side: 'seller' | 'customer';
        second_side: number;
        is_money: boolean;
    }

    function convertStatus(status: TransactionEntry_Status): TransactionRecord['status'] {
        switch (status) {
            case TransactionEntry_Status.CREATED:
                return 'created';
            case TransactionEntry_Status.APPROVED:
                return 'approved';
            case TransactionEntry_Status.REJECTED:
                return 'rejected';
            case TransactionEntry_Status.COMPLETED:
                return 'completed';
            default:
                return 'created';
        }
    }

    export async function createProductProposal(
        seller_account: number,
        customer_account: number,
        product: string,
        count: number,
        amount: number
    ): Promise<Response> {
        const req = CreateProductTransactionRequest.create({
            seller_account,
            customer_account,
            product,
            count,
            amount,
        });

        const encoded = CreateProductTransactionRequest.encode(req).finish();
        return util.wrapApiCall(() =>
            fetch('/api/transaction/product/create', {
                method: 'POST',
                headers: { 'Content-Type': 'application/octet-stream' },
                credentials: 'include',
                body: encoded,
            })
        );
    }

    export async function createMoneyProposal(
        seller_account: number,
        customer_account: number,
        amount: number
    ): Promise<Response> {
        const req = CreateMoneyTransactionRequest.create({
            seller_account,
            customer_account,
            amount,
        });

        const encoded = CreateMoneyTransactionRequest.encode(req).finish();
        return util.wrapApiCall(() =>
            fetch('/api/transaction/money/create', {
                method: 'POST',
                headers: { 'Content-Type': 'application/octet-stream' },
                credentials: 'include',
                body: encoded,
            })
        );
    }

    export async function getActiveProposals(account: number): Promise<TransactionRecord[]> {
        const req = ViewCurrentTransactionsRequest.create({ account });
        const encoded = ViewCurrentTransactionsRequest.encode(req).finish();

        const response = await util.wrapApiCall(() =>
            fetch('/api/transaction/view/current', {
                method: 'POST',
                headers: { 'Content-Type': 'application/octet-stream' },
                credentials: 'include',
                body: encoded,
            })
        );

        const buffer = new Uint8Array(await response.arrayBuffer());
        const decoded = CurrentTransactionsResponse.decode(buffer);

        return decoded.transactions.map((entry: CurrentTransactionEntry) => ({
            transaction_id: entry.transactionId,
            product: entry.product,
            count: entry.count,
            amount: entry.amount,
            status: 'created',
            updated_at: '',
            side: 'customer',
            second_side: entry.secondSide,
            is_money: false,
        }));
    }

    export async function getTransactionHistory(account: number): Promise<TransactionRecord[]> {
        const req = ViewHistoryRequest.create({ account });
        const encoded = ViewHistoryRequest.encode(req).finish();

        const response = await util.wrapApiCall(() =>
            fetch('/api/transaction/view/history', {
                method: 'POST',
                headers: { 'Content-Type': 'application/octet-stream' },
                credentials: 'include',
                body: encoded,
            })
        );

        const buffer = new Uint8Array(await response.arrayBuffer());
        const decoded = ViewTransactionsResponse.decode(buffer);

        return decoded.transactions.map((entry) => ({
            transaction_id: entry.transactionId,
            product: entry.product,
            count: entry.count,
            amount: entry.amount,
            status: convertStatus(entry.status),
            updated_at: entry.updatedAt,
            side: entry.side as 'seller' | 'customer',
            second_side: entry.secondSide,
            is_money: entry.isMoney,
        }));
    }

    export async function decideOnProposal(
        transaction_id: number,
        status: 'approved' | 'rejected'
    ): Promise<Response> {
        const req = DecideTransactionRequest.create({
            account: transaction_id,
            status,
        });

        const encoded = DecideTransactionRequest.encode(req).finish();
        return util.wrapApiCall(() =>
            fetch('/api/transaction/decide', {
                method: 'POST',
                headers: { 'Content-Type': 'application/octet-stream' },
                credentials: 'include',
                body: encoded,
            })
        );
    }
}

File: app/api/transactions/transaction.py
from datetime import datetime

import sqlalchemy as orm
from flask_login import login_required

from app.api import Blueprints
from app.api.helpers import local_datetime, protobufify, pythonify
from app.api.transactions.processor import Processor, complete_transaction
from app.codegen.hope import Response as APIResponse
from app.codegen.transaction import (
    CreateMoneyTransactionRequest,
    CreateProductTransactionRequest,
    CreateTransactionResponse,
    CurrentTransactionsRequest,
    CurrentTransactionsResponse,
    DecideOnTransactionRequest,
    ViewTransactionsRequest,
    ViewTransactionsResponse,
)
from app.codegen.types import Transaction as EntryProto
from app.codegen.types import TransactionStatus as TxStatus
from app.codegen.types import TransactionStatusReason
from app.context import AppContext
from app.models import Product, Transaction, TransactionStatus
from app.models.queries import wrap_crud_context


def make_status(status: TransactionStatusReason):
    return CreateTransactionResponse(status=status)


@Blueprints.transactions.route("/api/transaction/product/create", methods=["POST"])
@login_required
@pythonify(CreateProductTransactionRequest)
def new_proposal(ctx: AppContext, req: CreateProductTransactionRequest):
    products = ctx.database.session.scalars(
        orm.select(Product).filter_by(name=req.product)
    ).all()

    if len(products) != 1:
        return protobufify(
            APIResponse(
                create_transaction=make_status(TransactionStatusReason.MULTIPLE_PRODUCTS)
            )
        )

    product = products[0]
    with wrap_crud_context():
        tx = Transaction(
            product.id,
            req.customer_bank_account_id,
            req.seller_bank_account_id,
            req.count,
            req.amount,
            TransactionStatus.CREATED,
            datetime.now(),
            datetime.now(),
            "",
        )

        for check in [
            Processor.check_seller_not_customer,
            Processor.check_bounds,
            Processor.check_seller_exists,
            Processor.check_customer_exists,
        ]:
            status = check(ctx, tx) if 'exists' in check.__name__ else check(tx)
            if status != TransactionStatusReason.OK:
                return protobufify(
                    APIResponse(create_transaction=make_status(status))
                )

        ctx.database.session.add(tx)
        ctx.database.session.commit()

    return protobufify(
        APIResponse(create_product_transaction=make_status(TransactionStatusReason.OK))
    )


@Blueprints.transactions.route("/api/transaction/money/create", methods=["POST"])
@login_required
@pythonify(CreateMoneyTransactionRequest)
def new_money_proposal(ctx: AppContext, req: CreateMoneyTransactionRequest):
    with wrap_crud_context():
        tx = Transaction(
            ctx.config.money_product_id,
            req.customer_bank_account_id,
            req.seller_bank_account_id,
            req.amount,
            req.amount,
            TransactionStatus.CREATED,
            datetime.now(),
            datetime.now(),
            "",
        )

        result = Processor.process(ctx, tx, approved=True)
        if result != TransactionStatusReason.OK:
            return protobufify(
                APIResponse(create_transaction=make_status(result))
            )

        ctx.database.session.add(tx)
        ctx.database.session.commit()
        return protobufify(
            APIResponse(create_money_transaction=make_status(TransactionStatusReason.OK))
        )


@Blueprints.transactions.route("/api/transaction/current", methods=["POST"])
@login_required
@pythonify(CurrentTransactionsRequest)
def active_proposals(ctx: AppContext, req: CurrentTransactionsRequest):
    entries = ctx.database.session.execute(
        orm.select(Transaction, Product)
        .filter(
            orm.and_(
                Transaction.status == TransactionStatus.CREATED,
                Transaction.customer_bank_account_id == req.account,
            )
        )
        .join(Product, Product.id == Transaction.product_id)
    ).all()

    return protobufify(
        APIResponse(
            current_transactions=CurrentTransactionsResponse(
                transactions=[
                    EntryProto(
                        transaction_id=tx.id,
                        amount=tx.amount,
                        count=tx.count,
                        product=product.name,
                        seller_bank_account_id=tx.seller_bank_account_id,
                    )
                    for tx, product in entries
                ]
            )
        )
    )


@Blueprints.transactions.route("/api/transaction/history", methods=["POST"])
@login_required
@pythonify(ViewTransactionsRequest)
def view_proposal_history(ctx: AppContext, req: ViewTransactionsRequest):
    def fetch(for_seller: bool):
        column = Transaction.seller_bank_account_id if for_seller else Transaction.customer_bank_account_id
        return ctx.database.session.execute(
            orm.select(Transaction, Product)
            .filter(column == req.account)
            .join(Product, Product.id == Transaction.product_id)
        ).all()

    str_format = "%Y-%m-%d %H:%M"
    response = []
    for group, role in [(fetch(True), "seller"), (fetch(False), "customer")]:
        for tx, product in group:
            response.append(
                EntryProto(
                    transaction_id=tx.id,
                    amount=tx.amount,
                    count=tx.count,
                    product=product.name,
                    status=TxStatus.from_string(tx.status.upper()),
                    updated_at=local_datetime(ctx, tx.updated_at).strftime(str_format),
                    side=role,
                    is_money=(product.id == ctx.config.money_product_id),
                )
            )

    response.sort(
        key=lambda e: datetime.strptime(e.updated_at, str_format), reverse=True
    )
    return protobufify(
        APIResponse(
            view_transaction_history=ViewTransactionsResponse(transactions=response)
        )
    )


@Blueprints.transactions.route("/api/transaction/decide", methods=["POST"])
@login_required
@pythonify(DecideOnTransactionRequest)
def decide_on_proposal(ctx: AppContext, req: DecideOnTransactionRequest):
    result = complete_transaction(req.id, req.status)
    ctx.logger.info(f"transaction {req.id} decision={req.status}: {result}")

    return protobufify(APIResponse(decide_on_transaction=result))
